#165177127 Users should be able to edit their articles

reifred opened the pull request:
#### What does this PR do?
* Add an endpoint to update an article
#### Description of Task to be completed?
* Ensure only authenticated users edit articles
* Ensure only users who created an article can edit it
#### How should this be manually tested?
* After cloning this repo, cd into it and switch to this branch
* Start the development server using ```python manage.py runserver```

Given an existing article such as below
```
{
    "article": {
        "slug": "lorem-ipsum-dolor-old",
        "title": "Lorem ipsum dolor old",
        "description": "Lorem ipsum dolor sit amet, consectetur",
        "body": "Ask Question I added validate_mail and also installed",
        "createdAt": "2019-04-12 07:54:03.437416+00:00",
        "updatedAt": "2019-04-12 16:46:32.699055+00:00",
        "read_time": "1 min read",
        "favoritesCount": 0,
        "author": {
            "username": "reifred",
            "bio": "",
            "image": ""
        }
    }
}
```
* Using Postman test the endpoint below

Update using an article slug name in this case.
```
PATCH '/api/articles/lorem-ipsum-dolor-old/'
Body Request.
{
	"article":{
		"body": "blog body for post 1 changed"
	}
}
Expected Output
{
    "article": {
        "slug": "lorem-ipsum-dolor-old",
        "title": "Lorem ipsum dolor old",
        "description": "Lorem ipsum dolor sit amet, consectetur",
        "body": "blog body for post 1 changed",
        "createdAt": "2019-04-12 07:54:03.437416+00:00",
        "updatedAt": "2019-04-12 16:46:32.699055+00:00",
        "read_time": "1 min read",
        "favoritesCount": 0,
        "author": {
            "username": "reifred",
            "bio": "",
            "image": ""
        }
    }
}
```
In case an article slug given doesnot exist.
```
PATCH '/api/articles/blog-post-2/'
Request Body.
{
	"article":{
		"body": "blog body for post 1 changed"
	}
}
Expected Output
{
    "errors": {
        "detail": "Article does not exist in your articles."
    }
}
```
Error expected in case an author tries to edit another authors' article.
```
{
    "errors": {
        "detail": "You do not have permission to perform this action."
    }
}
```
#### What are the relevant pivotal tracker stories?
[#165177127](https://www.pivotaltracker.com/story/show/165177127)

RonKbS commented on authors/apps/articles/exceptions.py:
```diff
@@ -0,0 +1,5 @@
+
+class ArticleDoesNotExist(APIException):
+    status_code = 404
+    default_detail = "Requested Article is inexistent in your articles"
```
Please change `inexistent` to `does not exist`. It is simpler which is what we should always strive for when writing APIs.

RonKbS commented on authors/apps/articles/slug_creator.py:
```diff
@@ -0,0 +1,21 @@
+from django.template.defaultfilters import slugify
```
Please rename this file to something generic like `utilities.py` incase we need to add more functions that are required in more than one place. You could also move it to `authors/apps` such that we are able to access it across different apps.

RonKbS commented on authors/apps/articles/tests/test_data.py:
```diff
@@ -18,3 +18,6 @@
 }
+update_article_title = {
+    "title": "How to train my dragon",
+}
```
Ensure that you have new lines in all the files you have made changes to.

RonKbS reviewed (changes requested):
Great work @reifred.
Kindly implement the changes ASAP.

reifred commented:
> Great work @reifred.
> Kindly implement the changes ASAP.

Thanks for the feedback @RonKbS. Requested changes have been implemented

NicholusMuwonge commented on authors/apps/articles/models.py:
```diff
@@ -28,26 +27,10 @@ class Article(models.Model):
-        # the slag we created above already exists. If it does,
-        # make it unique by appending a hyphen and the extension
-        # to it.
-        while ModelClass.objects.filter(
```
Please help me understand your reason for removing this section that makes the slug unique in case its  duplicated?

reifred replied:
Thanks, @NicholusMuwonge for reviewing. The function was not removed but only its location was changed so that it can be reused

NicholusMuwonge commented on authors/apps/core/utilities.py:
```diff
@@ -0,0 +1,21 @@
+from django.template.defaultfilters import slugify
+
+
```
I think should be in a class for future usage

reifred replied:
I usually use classes to group related methods but in this case, I think a function is what I needed. How do you see?

NicholusMuwonge reviewed (changes requested):
let test it locally but the code looks okay just put the function in the core.utilities in a class

NicholusMuwonge commented:
This works fine locally. Thanks for the efforts @reifred

richien commented on authors/apps/articles/views.py:
```diff
@@ -1,27 +1,46 @@
+class UpdateArticleAPIView(UpdateAPIView):
+    queryset = Article.objects.select_related('author')
+    renderer_classes = (ArticleJSONRenderer,)
+    permission_classes = (IsAuthenticatedOrReadOnly,)
```
Hi @reifred, this is good work. Keep it up. I have tested the endpoint locally and was able to update an article. The only issues I have found so far is that;
1. Any authenticated user can edit any Article. Kindly restrict this to only the author of the Article.
2. I have not been able to test this new implementation of creating an Article because you have not yet updated this branch. Kindly do so, so that I can re-test.

Overall, this is really good and very insightful, thank you for the work.

reifred replied:
Thanks @richien, for the feedback. Kindly re-test again with the updated branch now.

richien replied:
Hi @reifred, thank you for implementing the feedback. I have tested and the route is working as expected. Great work.

RonKbS commented on authors/apps/articles/models.py:
```diff
@@ -17,41 +16,23 @@ class Article(models.Model):
         on_delete=models.CASCADE)
 
-    read_time = models.TextField()
+    read_time = models.CharField(max_length=255)
```
Kindly refrain from adding changes not related to a story you pick up.
If you feel this is a fix that needs addressing, get in touch with the TTL or PO to create a story for it.

Update: Looking through the rest of the code, is there a reason you've included  a restriction of 255 characters?

reifred replied:
Ohh, am sorry about this. This change came as a result when I was testing since at first I thought that read_time field was storing a body field during its calculation when determining time. So I decided to restrict its characters and use a body of characters greater than 255 characters to see whether I could get any errors. But later I figured out that its use was just to display the time of reading. Then mistakenly I commited it with my updates but it wasn't intentional to change it. But it's now fixed.

NicholusMuwonge reviewed (approved):
thanks for the nice work done.

RonKbS commented on authors/apps/articles/serializers.py:
```diff
@@ -4,40 +4,69 @@
+        return Article.objects.create(author=author, **validated_data)
 
-class ReadArticlesSerializer(serializers.ModelSerializer):
+    # This is function is to be removed after adding profile serializer
```
Has the profile serializer not been added with the recent PR merged?

RonKbS commented on authors/apps/articles/views.py:
```diff
@@ -1,55 +1,64 @@
-            status=status.HTTP_200_OK)
+    def update(self, request, slug, *args, **kwargs):
+        try:
+            article_toupdate = self.queryset.get(slug=slug)
```
Should `article_toupdate` not be `article_to_update`

RonKbS commented on authors/apps/articles/migrations/0001_initial.py:
```diff
@@ -1,6 +1,5 @@
-# Generated by Django 2.1.7 on 2019-04-05 07:52
```
Please explain the reason for the changes in this migration file and the steps you went through to introduce those changes

reifred replied:
I changed the migrations because at first, our Article model used the User instance as the foreign key and after getting the new updates of profile I had to change the user instance to Profile instance. But since the current migrations also were using the User instance I had to reset them so that they can use Profile instance instead. Without resetting them I would end up with errors for every operation I carried out since the migrations where expecting User instance and I was feeding in Profile instance. 
Then I used this link on how to reset them.
https://simpleisbetterthancomplex.com/tutorial/2016/07/26/how-to-reset-migrations.html.

RonKbS replied:
Thanks for the info.
However, when working with migrations it is better to generate a new file rather than making adjustments to an old one. This will ensure that we do not tamper with existing data in the database among other reasons.

Therefore, please remove any changes you've made to this file then run `python manage.py makemigrations` to generate a new migration file that you can push as part of your implementation.

reifred replied:
Thanks @RonKbS, at first I followed that implementation but I could result in errors. And I found the same issue raised here https://code.djangoproject.com/ticket/22319 were by it is claimed to be solved but some are getting problems with it and the simplest way to solve it was to reset migrations. 

Another solution here https://stackoverflow.com/questions/33496333/valueerror-related-model-uapp-model-cannot-be-resolved was to alter the dependencies but since I had changed from User instance to Profile instance still I was getting into errors.

RonKbS replied:
Kindly elaborate on the errors that were being generated following that implementation.

reifred replied:
It's the ValueError that happens on the line below whereby migrations ask for an instance that was changed to Profile instance.
```dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]
```
https://stackoverflow.com/questions/33496333/valueerror-related-model-uapp-model-cannot-be-resolved

RonKbS replied:
I just ran migrations successfully on your branch using that particular method and was successful (I reverted the changes in this migration file you that you had implemented).
<img width="1403" alt="Screenshot 2019-04-15 at 17 42 23" src="https://user-images.githubusercontent.com/40039818/56141786-29159180-5fa6-11e9-84e2-5b16bc6de4e1.png">

Can you list the exact steps you ran that resulted in the error being raised?

reifred replied:
@RonKbS Kindly run the tests with that new implementation because I too don't get any error in making and migrating migrations. But after running tests, this is the error that arises.
![error](https://user-images.githubusercontent.com/22131999/56143352-0f297e00-5fa9-11e9-994e-a405c1ea3f6c.PNG)

RonKbS commented on authors/apps/articles/migrations/0002_article_read_time.py:
```diff
@@ -1,19 +0,0 @@
-# Generated by Django 2.1.7 on 2019-04-08 15:49
```
Going forward, please refrain from tampering with migration files. Always generate a new one rather than making changes to an old one if you have changes in models.py files. This will make it easier to apply migrations without corrupting data if we were working with a database that contains actual data.

Return this file and generate a new file to which you'll add `('profiles', '0001_initial')` at the top of the dependencies list.
Also, undo all changes to `authors/apps/articles/migrations/0001_initial.py`

reifred committed:
feat(edit article): ensure users can edit their articles

- ensure only authenticated users edit articles
- ensure only users who created an article can edit it

[Finishes #165177127]
